OH2-205: Add code generation on Swagger

ArnaudFonzam opened the pull request:
See OH2-205 and OH2-209.

This pull request is to solve the problem of code genation that are false by default, In other to avoid special character when we generate the specification code. I have also change PageResponseDTO with Page because the name is too long.

Paired with https://github.com/informatici/openhospital-core/pull/1035.

**TODO**
- [x] now with swagger 3 there's no need to add "Bearer " in the swagger interface, so I would change the README too
- [x] now with swagger 3 the UI responds to http://localhost:8080/swagger-ui/index.html to update in the README
- [x] in swagger UI there are two endpoints /auth/login to be fixed (only one) as suggested

mwithi commented:
> I have also change PageResponseDTO with Page because the name is too long

I can see only this change. Also, there are conflicts with the develop branch: did you sync your develop before to make changes?

mwithi commented:
What about the "security scope" errors?

![image](https://github.com/informatici/openhospital-api/assets/2938553/fed8b7c1-062a-4205-8304-0ce8b9432594)

mwithi commented:
After some research, it seems to me that scopes are not applicable to `ApiKey` type, so to solve the "scope error", it should be enough to pass an empty array here:

```
	List<SecurityReference> defaultAuth() {
		return Arrays.asList(new SecurityReference("JWT", new AuthorizationScope[0]));
	}
```

After that, still some errors occur:

![image](https://github.com/informatici/openhospital-api/assets/2938553/9de9ae25-7ee9-457e-b93b-2e4b04d6edda)

ArnaudFonzam commented:
I have add the empty array in defaultAuth like you suggest but i have new error: 
![serror](https://github.com/informatici/openhospital-api/assets/101590821/63670c6f-c745-4f99-af99-0f8ae63cff78)

ArnaudFonzam commented:
The error about equivalent path error It about the new specification of OpenAPI that not allow the same path with different verb. look here : https://github.com/swagger-api/swagger-editor/issues/1677

mwithi commented:
> The error about equivalent path is cause by the new specification of OpenAPI that not allow the same path with different verb. look here : [swagger-api/swagger-editor#1677](https://github.com/swagger-api/swagger-editor/issues/1677)

I think the "equivalent path" is a symptom the API are not well organized and it could be a suggestion to adopt better practises.

What about the other errors?

ArnaudFonzam commented:
The last error is because in TherapyDTO we have array of date which not have a good reference. please why we have array of date in TherapyDTO
![therapy](https://github.com/informatici/openhospital-api/assets/101590821/b0876cd0-9835-41b2-8e1f-d3e179c9a025)

mwithi commented:
> The last error is because in TherapyDTO we have array of date which not have a good reference. please why we have array of date in TherapyDTO

It's the list of dates in which the patient should take the drugs

ArnaudFonzam commented:
I have try to solve error and i have that last error
![enerror](https://github.com/informatici/openhospital-api/assets/101590821/86d7399c-81a9-4587-bfc6-2552ef69ad1f)

ArnaudFonzam commented:
yes I am working on its

mwithi commented:
with latest version I can still see some "equivalent path" errors

![image](https://github.com/informatici/openhospital-api/assets/2938553/03a6ca3a-3ddb-47d7-9414-206060142808)

ArnaudFonzam commented:
> with latest version I can still see this error:
> 
> ![image](https://user-images.githubusercontent.com/2938553/250148290-b29122e6-b002-4d08-961c-c9486be4130d.png)
> 
> Moreover, I think the UI should be updated with the new API, so a new PR is needed on that side!
yes. but this error is genereted automatically  it is about the  response status of LoginController. and we can remove the line of error after generation

mwithi commented:
> yes. but this error is genereted automatically it is about the response status of LoginController. and we can remove the line of error after generation

is it related to https://github.com/informatici/openhospital-ui/pull/483 ?
Anyway I added under the screenshot "Moreover, I think the UI should be updated with the new API, so a new PR is needed on that side."

ArnaudFonzam commented:
> > yes. but this error is genereted automatically it is about the response status of LoginController. and we can remove the line of error after generation
> 
> is it related to [informatici/openhospital-ui#483](https://github.com/informatici/openhospital-ui/pull/483) ? Anyway I added under the screenshot "Moreover, I think the UI should be updated with the new API, so a new PR is needed on that side."

okay but the error appear because the status 200 don't have the description

> > yes. but this error is genereted automatically it is about the response status of LoginController. and we can remove the line of error after generation
> 
> is it related to [informatici/openhospital-ui#483](https://github.com/informatici/openhospital-ui/pull/483) ? Anyway I added under the screenshot "Moreover, I think the UI should be updated with the new API, so a new PR is needed on that side."

the error is not about this fond-end task. it is related to the logout api. the error appear because the 200 status don't have description. to solve it we can just remove the 200 status. 

![errL](https://github.com/informatici/openhospital-api/assets/101590821/0991c187-6884-4086-99cc-1beff312d6f1)

mwithi commented:
The new generated code is ok, but I think the UI is not yet ready for the changes in this PR. Please let's open a PR for OH2-205 also on front-end side

![image](https://github.com/informatici/openhospital-api/assets/2938553/55db8643-15b3-4470-bafc-47b0ce0885a5)

ArnaudFonzam commented:
> The new generated code is ok, but I think the UI is not yet ready for the changes in this PR. Please let's open a PR for [OH2-205](https://openhospital.atlassian.net/browse/OH2-205) also on front-end side
> 
> ![image](https://user-images.githubusercontent.com/2938553/250639413-55db8643-15b3-4470-bafc-47b0ce0885a5.png)

okay but I think we can close this PR because so that I can continous with new task

mwithi commented:
> okay but I think we can close this PR so that I can continous with new task. I have update the spec file and I will share it with steve in other to update the api route on the front end side

What I mean is that if we merge this, the front-end UI is automatically broken and we don't know for how long. I will not suggest this way of doing things... when the PR will be ready on the UI side, we will merge them together.

You can continue with other tasks starting from develop, changes in this branch are not affecting the whole codebase, you will be able to rebase your work onto develop after the merge.

ArnaudFonzam commented:
> > okay but I think we can close this PR so that I can continous with new task. I have update the spec file and I will share it with steve in other to update the api route on the front end side
> 
> What I mean is that if we merge this, the front-end UI is automatically broken and we don't know for how long. I will not suggest this way of doing things... when the PR will be ready on the UI side, we will merge them together.
> 
> You can continue with other tasks starting from develop, changes in this branch are not affecting the whole codebase, you will be able to rebase your work onto develop after the merge.

Okay so I will create the task and attach the spec file on its. and assign it to steve so that he will align the front-end with the backend.

mwithi commented on src/main/java/org/isf/shared/GenericMapper.java:
```diff
@@ -89,7 +89,8 @@ public PageInfoDTO setParameterPageInfo(PageInfo pageInfo) {
 		pageInfoDTO.setSize(pageInfo.getSize());
-		pageInfoDTO.setTotalCount(pageInfo.getTotalCount());
+		pageInfoDTO.setTotalNbOfElements(pageInfo.getTotalNbOfElements());
+		pageInfoDTO.setTotalPage(pageInfo.getTotalPage());
```
this seems to be the right method name, so far...

```suggestion
		pageInfoDTO.setTotalPage(pageInfo.getTotalPages());
```

mwithi commented on src/main/java/org/isf/security/LoginApi.java:
```diff
@@ -27,34 +27,35 @@
-@Api("Login")
+
 @RestController
+@Tag(name = "Login")
```
Move `logout` endpoint to appear under the same `login` tag

```suggestion
@Tag(name = "Auth")
```

mwithi commented on src/main/java/org/isf/security/LoginApi.java:
To remove the login controller because already in defined in org/isf/login/rest/LoginController.java

ArnaudFonzam replied:
if I have understand your suggestion is to remove login enpoind in LoginApi and Rename the LoginApi to LogoutApi. also Delete LoginResponse in the same package and use org/isf/login/dto/LoginResponse.java on OHSimpleUrlAuthenticationSuccessHandler. please is it ?

mwithi replied:
yes

ArnaudFonzam replied:
okay

mwithi reviewed (commented):
Some comments

mwithi commented on pom.xml:
Let's consider upgrade to:

```
    <groupId>org.isf</groupId>
    <artifactId>openhospital-api</artifactId>
    <version>0.1.0</version>
```

ArnaudFonzam replied:
okay

mwithi reviewed (changes requested):
Please review the comments

mwithi commented on src/main/java/org/isf/ward/rest/WardController.java:
```diff
@@ -43,11 +43,12 @@
-@RestController
-@Api(value = "/wards", produces = MediaType.APPLICATION_JSON_VALUE, authorizations = {@Authorization(value="apiKey")})
+@RestController(value = "/wards")
+@Tag(name = "Wards")
```
Please consider to add this to `application.properties.dist`

```
### Swagger-UI (info)
api.host=localhost:8080
api.protocol=http
springdoc.swagger-ui.tagsSorter=alpha <--- for tag sorting in the swagger UI
```

ArnaudFonzam replied:
okay

mwithi commented on src/main/java/org/isf/login/rest/LoginController.java:
```diff
@@ -46,10 +46,12 @@
-@RestController
-@Api(value = "/auth", produces = MediaType.APPLICATION_JSON_VALUE)
+@RestController(value = "/auth")
+@Tag(name = "Auth")
```
Move `login` endpoint to appear under the same `Login` tag

```suggestion
@Tag(name = "Login")
```

dbmalkovsky commented on src/main/java/org/isf/lab/rest/LaboratoryController.java:
```diff
@@ -549,41 +552,81 @@ public ResponseEntity<List<String>> getMaterials() throws OHServiceException {
+		List<Laboratory> labList = new ArrayList<>();
+		if (paged) {
+			if (!status.equals("")) {
+				if(examName.equals("")) {
```
```suggestion
				if (examName.equals("")) {
```

dbmalkovsky commented on src/main/java/org/isf/lab/rest/LaboratoryController.java:
```diff
@@ -549,41 +552,81 @@ public ResponseEntity<List<String>> getMaterials() throws OHServiceException {
+	                    .stream().filter(lab -> lab.getStatus().equalsIgnoreCase(status)).collect(Collectors.toList());
+				laboratoryPageable.setData(labList);
+			} else {
+				if(examName.equals("")) {
```
```suggestion
				if (examName.equals("")) {
```

dbmalkovsky commented on src/main/java/org/isf/lab/rest/LaboratoryController.java:
```diff
@@ -549,41 +552,81 @@ public ResponseEntity<List<String>> getMaterials() throws OHServiceException {
+	                    .stream().filter(lab -> lab.getStatus().equalsIgnoreCase(status)).collect(Collectors.toList());
+				laboratoryPageable.setData(labList);
+			} else {
+				if(examName.equals("")) {
```
```suggestion
				if (examName.equals("")) {
```

mwithi commented on rsc/application.properties.dist:
```diff
@@ -16,6 +16,8 @@ spring.pid.file=OH_API_PID
 ### Swagger-UI (info)
 api.host=localhost:8080
 api.protocol=http
+springdoc.swagger-ui.tagsSorter=alpha <--- for tag sorting in the swagger UI
+springdoc.swagger-ui.doc-expansion=none <--- for tags sections collapsed by default
```
```suggestion
springdoc.swagger-ui.tagsSorter=alpha
springdoc.swagger-ui.doc-expansion=none
```

mwithi commented on src/main/java/org/isf/security/LogoutApi.java:
```diff
@@ -21,40 +21,29 @@
-    void login(@Valid @RequestParam String username,  @RequestParam String password) {
-    	 throw new IllegalStateException("Add Spring Security to handle authentication");
-    }
+@Tag(name = "Logout")
```
```suggestion
@Tag(name = "Login")
```

So in Swagger the two endpoints are listed under the same tag:

![image](https://github.com/informatici/openhospital-api/assets/2938553/298a45db-b415-4d92-8c86-0ffac7acc72d)

instead of

![image](https://github.com/informatici/openhospital-api/assets/2938553/978d3894-0591-4c1e-aa55-1fdd2dbd5447)

mwithi replied:
Also please review (and accept if ok) my PR https://github.com/intesys/openhospital-api/pull/77

mwithi commented on src/main/java/org/isf/security/LogoutApi.java:
```diff
@@ -21,40 +21,29 @@
-    void login(@Valid @RequestParam String username,  @RequestParam String password) {
-    	 throw new IllegalStateException("Add Spring Security to handle authentication");
-    }
+@Tag(name = "Logout")
```
Why you changed and then reverted the change? This is the one suggested initially 

```suggestion
@Tag(name = "Login")
```
